sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c

```diff
@@ -0,0 +1,485 @@
+            memcpy(nce->l2addr, tl2ao + 1, l2addr_len);
+        }
+        else {
+            nce->l2addr_len = 0;
+        }
+        if (_sflag_set((ndp_nbr_adv_t *)icmpv6)) {
+            _set_reachable(iface, nce);
+        }
+        else if ((icmpv6->type != ICMPV6_NBR_ADV) ||
+                 !_sflag_set((ndp_nbr_adv_t *)icmpv6) ||
+                 (!nce_was_incomplete &&
+                  _tl2ao_changes_nce(nce, tl2ao, iface, l2addr_len))) {
+            DEBUG("nib: Set %s%%%u to STALE\n",
+                  ipv6_addr_to_str(addr_str, &nce->ipv6, sizeof(addr_str)),
+                  iface);
+            evtimer_del(&_nib_evtimer, &nce->nud_timeout.event);
+            _set_nud_state(nce, GNRC_IPV6_NIB_NC_INFO_NUD_STATE_STALE);
+        }
+        if (_oflag_set((ndp_nbr_adv_t *)icmpv6) ||
+            ((icmpv6->type == ICMPV6_NBR_ADV) && nce_was_incomplete)) {
+            if (_rflag_set((ndp_nbr_adv_t *)icmpv6)) {
+                nce->info |= GNRC_IPV6_NIB_NC_INFO_IS_ROUTER;
+            }
+            else {
+                nce->info &= ~GNRC_IPV6_NIB_NC_INFO_IS_ROUTER;
+            }
+        }
+#if GNRC_IPV6_NIB_CONF_QUEUE_PKT && MODULE_GNRC_IPV6
+        /* send queued packets */
+        gnrc_pktqueue_t *ptr;
+        DEBUG("nib: Sending queued packets\n");
+        while ((ptr = gnrc_pktqueue_remove_head(&nce->pktqueue)) != NULL) {
+            if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_IPV6,
+                                           GNRC_NETREG_DEMUX_CTX_ALL,
+                                           ptr->pkt)) {
+                DEBUG("nib: No receivers for packet\n");
+                gnrc_pktbuf_release_error(ptr->pkt, EBADF);
+            }
+            ptr->pkt = NULL;
+        }
+#endif  /* GNRC_IPV6_NIB_CONF_QUEUE_PKT */
+        if ((icmpv6->type == ICMPV6_NBR_ADV) &&
+            !_sflag_set((ndp_nbr_adv_t *)icmpv6) &&
+            (_get_nud_state(nce) == GNRC_IPV6_NIB_NC_INFO_NUD_STATE_REACHABLE) &&
+            _tl2ao_changes_nce(nce, tl2ao, iface, l2addr_len)) {
+            evtimer_del(&_nib_evtimer, &nce->nud_timeout.event);
+            _set_nud_state(nce, GNRC_IPV6_NIB_NC_INFO_NUD_STATE_STALE);
+        }
+    }
+    else if ((icmpv6->type == ICMPV6_NBR_ADV) &&
+             (_get_nud_state(nce) != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_INCOMPLETE) &&
+             (_get_nud_state(nce) != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_UNMANAGED) &&
+             _sflag_set((ndp_nbr_adv_t *)icmpv6) &&
+             !_tl2ao_changes_nce(nce, tl2ao, iface, l2addr_len)) {
+        _set_reachable(iface, nce);
+    }
+}
+
+void _set_reachable(unsigned iface, _nib_onl_entry_t *nce)
+{
+    _nib_iface_t *nib_netif = _nib_iface_get(iface);
+
+    /* reachable timeout wasn't initialized yet */
+    if (nib_netif->reach_time == 0) {
+        /* (re-)initialize PNRG with current system time as seed to get
+         * better randomization in reachable time */
+        random_init(xtimer_now_usec());
```

Comment:
There's no need to re-seed the PRNG **iff** the start seed is distinct on all nodes. And the latter is what we should aim for.

Reply:
And btw, you break here the PRNG for all applications that use this random interface with the same PRNG state.

Reply:
> There's no need to re-seed the PRNG iff the start seed is distinct on all nodes. And the latter is what we should aim for.

But it is not the case now. Also: this is a very standard approach to get a distinct random number.

> And btw, you break here the PRNG for all applications that use this random interface with the same PRNG state.

If re-seeding breaks your PNRG then your PNRG is broken

Reply:
Sorry, let me rephrase:

You are using the **system**/**global** PRNG state here. The moment you re-seed, the PRNG for all applications that make use of the **system**/**global** state do not produce *deterministic* results anymore. A PRNG is supposed to yield deterministic results.

Oh and btw: `random_init()` is **not** thread-safe. So please let the seeding be performed at auto_init() once, like it's the case now, or provide a separate PRNG state for NIB that can be re-seeded without global effects.

Reply:
>or provide a separate PRNG state for NIB that can be re-seeded without global effects.

and even then I doubt that re-seeding provides "better randomization".

Reply:
> and even then I doubt that re-seeding provides "better randomization".

Why are you doubting that? Currently the PNRG is *always* seeded with 0, there are plans to use the `luid` as seed (making it different but deterministic for different nodes). This reseed uses the reception time of the *first received* solicited NA from all neighbors, which is always **naturally** random, due to the different air times of both NS and NA, so it is by definition "better randomization" than using the PNRG with a deterministic seed.

Reply:
What about the other points?
1. `random_init()` is not thread-safe
2. you modify the global PRNG state which is potentially used by more thread than just the network stack

I propose that you factor out this re-seeding procedure into another PR and we continue the discussion there, so that we can continue here?

Reply:
Fine...

Reply:
Done